File: Blockpass/RSKeychain.m
//
//  RSKeychain.m
//
//  Created by daniel on 4/23/09.
//  Copyright 2009 Daniel Jalkut. All rights reserved.
//

#import "RSKeychain.h"
#import <Security/Security.h>

@implementation RSKeychain

+ (void) setSecureBytes:(const void *)bytePointer length:(NSUInteger)byteLength forItemName:(NSString*)keychainItemName accountName:(NSString*)accountName
{
    const char* itemNameUTF8String = [keychainItemName UTF8String];
	
	if (itemNameUTF8String != nil)
	{
		SecKeychainRef defaultKeychain;
		if (SecKeychainCopyDefault (&defaultKeychain) == noErr)
		{
			BOOL shouldAddNewItem = YES;

			const char *accountUTF8String = [accountName UTF8String];
		
			// Update an existing item if we have one
			SecKeychainItemRef itemRef = nil;
			OSErr err = SecKeychainFindGenericPassword (nil, strlen(itemNameUTF8String), itemNameUTF8String,
														strlen(accountUTF8String), accountUTF8String,
													NULL, NULL, &itemRef);
			if (err == noErr && itemRef)
			{		
				err = SecKeychainItemModifyContent (itemRef, NULL, byteLength, bytePointer);
				if (err == noErr)
				{
					shouldAddNewItem = NO;
				}
				
				CFRelease(itemRef);
			}
			
			if (shouldAddNewItem)
			{
				SecKeychainAddGenericPassword (defaultKeychain, strlen(itemNameUTF8String),
					itemNameUTF8String, strlen(accountUTF8String), accountUTF8String,
					byteLength, bytePointer, nil);
			}
			
			CFRelease(defaultKeychain);
		}
	}
}

+ (NSData*) secureDataForItemName:(NSString*)keychainItemName accountName:(NSString*)accountName
{
	NSData* foundData = nil;
	const char *accountNameUTF8String = [accountName UTF8String];
    const char* keychainItemNameUTF8String = [keychainItemName UTF8String];
	OSErr err = noErr;
	
	if (keychainItemNameUTF8String != nil)
	{
		UInt32 secureDataLength = 0;
		void *secureDataBuffer = NULL;
		
		err = SecKeychainFindGenericPassword (nil, strlen (keychainItemNameUTF8String),
			keychainItemNameUTF8String, strlen (accountNameUTF8String), accountNameUTF8String,
			&secureDataLength, (void **) &secureDataBuffer, nil);
		
		if (err == noErr)
		{
			foundData = [NSData dataWithBytes:(const void *)secureDataBuffer length:secureDataLength];

			SecKeychainItemFreeContent(NULL, secureDataBuffer);		
		}
		else
		{
			if (err != errSecItemNotFound)
			{
				NSLog(@"Got error %d trying to access password in keychain for %@/%@", err, keychainItemName, accountName);
			}
		}
	}
	
	return foundData;
}

+ (void) setSecureData:(NSData*)newData forItemName:(NSString*)keychainItemName accountName:(NSString*)accountName
{
	[self setSecureBytes:[newData bytes] length:[newData length] forItemName:keychainItemName accountName:accountName];
}

+ (void) setSecureString:(NSString*)passwordString forItemName:(NSString*)keychainItemName accountName:(NSString*)accountName
{
	const char *passwordUTF8String = [passwordString UTF8String];

	[self setSecureBytes:passwordUTF8String length:strlen(passwordUTF8String) forItemName:keychainItemName accountName:accountName];
}

+ (NSString*) secureStringForItemName:(NSString*)keychainItemName accountName:(NSString*)accountName
{
	NSData* passwordItemData = [self secureDataForItemName:keychainItemName accountName:accountName];
	NSString *foundPassword = [[NSString alloc] initWithData:passwordItemData encoding:NSUTF8StringEncoding];

	return foundPassword;
}


@end

File: Blockpass/RSKeychain.swift
//
//  RSKeychain.swift
//  Blockpass
//
//  Created by Daniel Jalkut on 8/31/18.
//  Copyright © 2018 Daniel Jalkut. All rights reserved.
//

import Foundation

struct RSKeychain {
	internal static func setSecureBytes(_ bytePointer: UnsafeRawPointer!, length byteLength: UInt, forItemName keychainItemName: String, accountName: String) {

		var defaultKeychain: SecKeychain! = nil
		if SecKeychainCopyDefault(&defaultKeychain) == noErr {
			var shouldAddNewItem = true

			let keychainItemNameCString = (keychainItemName as NSString).utf8String
			let keychainItemNameLength = UInt32(strlen(keychainItemNameCString))

			let accountNameCString = (accountName as NSString).utf8String
			let accountNameLength = UInt32(strlen(accountNameCString))

			// Update an existing item if we have one
			var keychainItemRef: SecKeychainItem? = nil
			if (SecKeychainFindGenericPassword(nil, keychainItemNameLength, keychainItemNameCString, accountNameLength, accountNameCString, nil, nil, &keychainItemRef) == noErr),
				let keychainItemRef = keychainItemRef {
				if SecKeychainItemModifyContent(keychainItemRef, nil, UInt32(byteLength), bytePointer) == noErr {
					shouldAddNewItem = false
				}
			}

			if shouldAddNewItem {
				SecKeychainAddGenericPassword(defaultKeychain, keychainItemNameLength, keychainItemNameCString, accountNameLength, accountNameCString, UInt32(byteLength), bytePointer, nil)
			}
		}
	}

	public static func secureData(forItemName keychainItemName: String, accountName: String) -> Data? {
		var foundData: Data? = nil

		let keychainItemNameCString = (keychainItemName as NSString).utf8String
		let keychainItemNameLength = UInt32(strlen(keychainItemNameCString))

		let accountNameCString = (accountName as NSString).utf8String
		let accountNameLength = UInt32(strlen(accountNameCString))

		var secureDataLength: UInt32 = 0
		var secureData: UnsafeMutableRawPointer? = nil
		let secureErr = SecKeychainFindGenericPassword(nil, keychainItemNameLength, keychainItemNameCString, accountNameLength, accountNameCString, &secureDataLength, &secureData, nil)
		if secureErr == noErr,
			let secureData = secureData
		{
			foundData = Data(bytes: secureData, count: Int(secureDataLength))
		}
		else {
			if secureErr != errSecItemNotFound {
				NSLog("Got error %d trying to access password in keychain for %@/%@", secureErr, keychainItemName, accountName);
			}
		}
		return foundData
	}

	public static func setSecureString(_ newString: String, forItemName keychainItemName: String, accountName: String) {
		let secureString = (newString as NSString).utf8String
		self.setSecureBytes(secureString, length: UInt(strlen(secureString)), forItemName: keychainItemName, accountName: accountName)
	}

	public static func secureString(forItemName keychainItemName: String, accountName: String) -> String? {
		guard
			let secureData = self.secureData(forItemName: keychainItemName, accountName: accountName),
			let secureString = String(data: secureData, encoding: .utf8)
		else {
			return nil
		}

		return secureString
	}
}
